openebs/cmd/openebs-provisioner/main.go

```
@@ -0,0 +1,112 @@
+	CASTemplateFeatureGateENVK = "OPENEBS_IO_CAS_TEMPLATE_FEATURE_GATE"
+)
+
+func main() {
+	syscall.Umask(0)
+
+	flag.Parse()
+	flag.Set("logtostderr", "true")
+	var (
+		config     *rest.Config
+		err        error
+		k8sMaster  = mayav1.K8sMasterENV()
+		kubeConfig = mayav1.KubeConfigENV()
+	)
+	if len(k8sMaster) != 0 || len(kubeConfig) != 0 {
+		glog.Infof("Build client config using k8s Master's Address: '%s' or Kubeconfig: '%s' \n", k8sMaster, kubeConfig)
+		config, err = clientcmd.BuildConfigFromFlags(k8sMaster, kubeConfig)
+	} else {
+		// Create an InClusterConfig and use it to create a client for the controller
+		// to use to communicate with Kubernetes
+		config, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		glog.Errorf("Failed to create config: %v", err)
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		glog.Errorf("Failed to create client: %v", err)
+	}
+
+	// The controller needs to know what the server version is because out-of-tree
+	// provisioners aren't officially supported until 1.5
+	serverVersion, err := clientset.Discovery().ServerVersion()
+	if err != nil {
+		glog.Errorf("Error getting server version: %v", err)
+	}
+
+	// Create the provisioner: it implements the Provisioner interface expected by
+	// the controller
+	var openEBSProvisioner controller.Provisioner
+	if CASTemplateFeatureGate() {
+		glog.Infof("Using CAS template feature gate for volume provisioning")
+		openEBSProvisioner = provisioner.NewOpenEBSCASProvisioner(clientset)
+	} else {
+		openEBSProvisioner = provisioner.NewOpenEBSProvisioner(clientset)
+	}
+	if openEBSProvisioner != nil {
+		// Start the provision controller which will dynamically provision OpenEBS PVs
+		pc := controller.NewProvisionController(
+			clientset,
+			provisionerName,
+			openEBSProvisioner,
+			serverVersion.GitVersion)
+
+		// Run starts all of controller's control loops
+		pc.Run(wait.NeverStop)
+	} else {
+		os.Exit(1) //Exit if provisioner not created.
```

if we can exit when provisioner could not be created. Why can we not exit if the kubeconfig & kube versions could not be determined. this will lead to crashloop but then that's ok IMO.

I still need to understand who will trigger the re-attempts as you have mentioned earlier @prateekpandey14

---

done, handled the silent errors.